crypto/x509/v3_ncons.c

```diff
@@ -782,50 +783,56 @@ static int nc_email(ASN1_IA5STRING *eml, ASN1_IA5STRING *base)
 static int nc_uri(ASN1_IA5STRING *uri, ASN1_IA5STRING *base)
 {
     const char *baseptr = (char *)base->data;
-    const char *hostptr = (char *)uri->data;
-    const char *p = ia5memchr(uri, (char *)uri->data, ':');
+    char *uri_copy;
+    char *scheme;
+    char *host;
     int hostlen;
+    int ret;
 
-    /* Check for foo:// and skip past it */
-    if (p == NULL
-            || IA5_OFFSET_LEN(uri, p) < 3
-            || p[1] != '/'
-            || p[2] != '/')
-        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
-    hostptr = p + 3;
-
-    /* Determine length of hostname part of URI */
+    if ((uri_copy = OPENSSL_strndup((const char *) uri->data, uri->length)) == NULL)
+        return X509_V_ERR_UNSPECIFIED;
 
-    /* Look for a port indicator as end of hostname first */
+    if (!OSSL_parse_url(uri_copy, &scheme, NULL, &host, NULL, NULL, NULL, NULL, NULL)) {
+        OPENSSL_free(uri_copy);
+        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
+    }
 
-    p = ia5memchr(uri, hostptr, ':');
-    /* Otherwise look for trailing slash */
-    if (p == NULL)
-        p = ia5memchr(uri, hostptr, '/');
+    /* Make sure the scheme is there */
+    if (scheme == NULL) {
+        fprintf(stderr, "x509: missing scheme in URI: %s\n", uri_copy);
+        OPENSSL_free(uri_copy);
+        ret = X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
+        goto end;
+    }
```

Comment:
Note that `scheme == NULL` cannot happen on success.
Also since you anyway do not use the scheme, you can leave out all this.

Reply:
Hmm, yet maybe it is required to check that the scheme has at least 3 chars, as was done before?

Reply:
> Note that `scheme == NULL` cannot happen on success. 

How so? On entry to `OSSL_parse_url` the scheme will be initialised to `NULL` and if the URI is missing `://`, the function may return success.

Reply:
> Hmm, yet maybe it is required to check that the scheme has at least 3 chars, as was done before?

The previous code merely skipped over the scheme, it did not actually check whether it was at least 3 characters.
The only thing it was doing was to skip over `://`, to get to the hostname part. Current code is equivalent in that regard.